mitodl/unified-ecommerce
# Adjust routing, fix logout, move session secret key to own variable

### What are the relevant tickets?

mitodl/hq#5656

### Description (What does it do?)

The main thing this does is moves most of the routes _out_ of the configuration that requires OIDC. The app maintains a Django session, so that is sufficient, and having all the API routes run through the OIDC plugin mean that we end up with CORS errors and other weirdness when it invariably tries to punt the request through to Keycloak. 

The routes that are still set to use the OIDC plugin are:
- `/cart` - the cart test mule
- `/admin` - the Django Admin
- `/establish_ession` - establishes the session for a user (the frontend redirects here if you're not logged in)

The rest are located in the unauthenticated route. APISIX will not try to resolve a Keycloak session, which will leave the app to use the existing Django session if there is one. (Unauthenticated in this case is from the viewpoint of APISIX.)

This also fixes a couple other things:
- Now setting a session secret - this was a thing that should have been set prior. The secret can be set via `APISIX_SESSION_SECRET_KEY` in the `.env` file and should have a reasonable default.
- Updated the OIDC config to send the user to the `UE_LOGOUT_URL` setting after logging out.
- Fixes logout so that it actually works now
   - Added an explicit route for `/logout` so that the OIDC plugin can intercept it. (Otherwise, the path doesn't match any known routes.) 
   - Added a new route to rewrite `/logout/` to `/logout` so both of these work.

### How can this be tested?

Set `UE_LOGOUT_URL` to a reasonable value: `UE_LOGOUT_URL=http://ue.odl.local:9080/auth/logout/`
This route specifically will clear the Django session.

Optionally: set `APISIX_SESSION_SECRET_KEY` to a reasonable value. It does have a default, so you don't _have_ to specify it. If you do set it, the key needs to be a non-numeric string of at least 16 characters. (In other words, if you just provide numbers APISIX will treat it as a number and will fail to parse the routes, and it will return 404 for all URLs.) 

1. Go to an API route that requires auth. (The basket list one is a good choice.) You shouldn't get anything back.
2. Go to an API route that _doesn't_ require auth. (The product list one is a good choice.) You should get data.
3. Set up a session. An easy way to do this is to just go to the cart test mule or to the Django Admin; both of these should redirect you through Keycloak. 
4. Check the "me" endpoint: `/api/v0/users/me/` You should see your account info.
5. Check other API routes. You should be able to get data from the authenticated and unauthenticated routes.
6. Go to `/logout` or `/logout/`. You should be redirect to whatever you have set for `UE_LOGOUT_URL` in your `.env`. 

### Additional Notes

Some notes for later use: APISIX's OIDC plugin will intercept a URI to handle logout properly (through the IdP). This URI is configurable, but doesn't support wildcards. Additionally, the URI needs to be in the URI set for the route that uses the OIDC plugin - it won't add it to the list. This is not necessarily noted anywhere so it was sort of confusing when it didn't work initially.

We may be able to continue to route the API requests through OIDC without forcing auth when the user doesn't have an APISIX session. This would allow us to capture account data changes - if the user adjusts their profile outside of UE, hitting the API would refresh their profile within UE too (which is how it works without these changes). However, the session is separate between routes I think so this may be more trouble than it's worth. We may instead want to use SCIM to sync user profile data into UE, like we do(?) elsewhere.
